### PR TITLE
8.0 sale order revision

### PR DESCRIPTION
**Pull request by gurneyalex**

### CODE REVIEW

File: sale_order_revision/README.rst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+Revisions for sale orders (and quotations)
+==========================================
+
+On cancelled orders, you can click on the "New copy of Quotation". This will
+create a new revision of the quotation, with the same base number and a
+'-revno' suffix appended. A message is added in the chatter saying that a new
+revision was created.
+
+In the form view, a new tab is added that lists the previous revisions, with
+the date of their were made obsolete and the user who performed the action.
```
Comment (pedrobaeza) [inline sale_order_revision/README.rst]: the date when it was made ...

File: sale_order_revision/i18n/en.po [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# Translation of Odoo Server.
```
Comment (pedrobaeza) [inline sale_order_revision/i18n/en.po]: File not needed.

File: sale_order_revision/model/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    def create(self, values):
+        if 'unrevisioned_name' not in values:
+            if values.get('name', '/') == '/':
+                seq = self.env['ir.sequence']
+                values['name'] = seq.next_by_code('sale.order') or '/'
+            values['unrevisioned_name'] = values['name']
+        return super(sale_order, self).create(values)
+
+    def init(self, cr):
+        cr.execute('UPDATE sale_order '
```
Comment (pedrobaeza) [inline sale_order_revision/model/sale_order.py]: This should be done as a post_init_hook.

File: sale_order_revision/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    "depends": ['sale'],
+    "data": [
+        'view/sale_order.xml',
+        ],
+    "test": [
+        'test/sale_order.yml',
+        ],
+    "active": False,
+    "installable": True,
+    "post_init_hook": 'populate_unrevisoned_name',
```
Comment (eLBati) [inline sale_order_revision/__openerp__.py]: populate_unrevisoned_name --> populate_unrevisioned_name

File: sale_order_revision/test/sale_order.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+     self.action_cancel(cr, uid, [ref('sale_order_1')])
+-
+  I create a new revision
+-
+   !python {model: sale.order}: |
+     action = self.copy_quotation(cr, uid, [ref('sale_order_1')])
+     assert action['res_id'] == ref('sale_order_1')
+     new_so = self.browse(cr, uid, ref('sale_order_1'))
+     assert new_so.old_revision_ids, "Old revisions not set"
+     assert all(so.active == False for so in new_so.old_revision_ids)
```
Comment (eLBati) [inline sale_order_revision/test/sale_order.yml:22]: Could you just add a check about `name` and `revision_number`?

File: sale_order_revision/model/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    @api.model
+    def create(self, values):
+        if 'unrevisioned_name' not in values:
+            if values.get('name', '/') == '/':
+                seq = self.env['ir.sequence']
+                values['name'] = seq.next_by_code('sale.order') or '/'
+            values['unrevisioned_name'] = values['name']
+        return super(sale_order, self).create(values)
+
```
Comment (eLBati) [inline sale_order_revision/model/sale_order.py]: travis says

```
./sale_order_revision/model/sale_order.py:97:1: W391 blank line at end of file
```

**Comment by pedrobaeza** [comment]
@gurneyalex, is runbot working for this branch? Can you restart build then to try directly? Is there any chance that people can restart builds without specific permissions as on Odoo runbot? This is because janitor is going to kill old (and not so old instances), and there's no possibility then to make a functional review of an old PR.

**Comment by pedrobaeza** [comment]
:+1: when @eLBati's comments are honoured.

File: sale_order_revision/view/sale_order.xml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+                        <field name="old_revision_ids" >
+                            <tree>
+                                <field name='name'/>
+                                <field name='create_date' string="Superseeded on"/>
+                                <field name='create_uid' string="Superseeded by"/>
+                                <field name='state' invisible='1'/>
+                            </tree>
+                        </field>
+                        <group>
+                            <field name="current_revision_id"/>
```
Comment (pedrobaeza) [inline sale_order_revision/view/sale_order.xml:20]: This field is now meaningless in the view, because you only see the current revision on the list.

**Comment by gurneyalex** [comment]
@pedrobaeza @eLBati updated the PR with your remarks

**Comment by pedrobaeza** [comment]
Still 2 comments left: remove en.po file and remove current_revision_id field from view

File: sale_order_revision/view/sale_order.xml (hunk 7 above)
Comment (gurneyalex) [inline sale_order_revision/view/sale_order.xml:20]: I think it is still useful to have, when your are looking at a previous revision.

Comment (pedrobaeza) [inline sale_order_revision/view/sale_order.xml:20]: But you always look at this revision through the current one, doesn't it?

Comment (gurneyalex) [inline sale_order_revision/view/sale_order.xml:20]: there is always the possibility of using advanced search. 

I updated the view to only display the field if the SO is not active (and hence not the current revision).

**Comment by pedrobaeza** [comment]
:+1:

**Comment by eLBati** [comment]
thanks :+1:

**Pull request merged**